### PR TITLE
Duplicated oasis get calls

### LINKED ISSUE #816: Water levels time update issue
**Describe the bug**
The bug was discovered in a configuration having wind coupling and water levels read from file. In the ww3_shel program there is a call to the W3WAVE routine, and the time parameter of this routine specifying until what time we have to integrate is set after reading the different input fields in sequence. 

Water levels are read before winds. This field is read from a file, and it is not interpolated. Therefore, the 'ending time' for W3WAVE is set to the initial time of the run. After this, winds are read, and the 'ending time' for W3WAVE should be updated to be the next coupling time step. The problem is that this 'ending time' is different to the ending time set by the water levels and the difference is negative (DTTST<0 in the ww3_shel program), so the result is that W3WAVE is called with and 'ending time' equal to the initial time of the run. Within this W3WAVE routine, the OASIS put/get routines are called, and nothing is done because the run time is not equal to the coupling time, which causes an inconsistency with the rest of the models WWIII is coupled to, leading OASIS to crash.

In conclusion, the error happens because the driving data is read twice in the first time step. Therefore, the oasis get routines are called twice for the first time step, which causes an error in OASIS as they should only be called once.

One possible solution would be interpolating the water levels by default, but this greatly increases the computation time because the grid has to be transformed each time step. Another solution would be not updating the 'ending time' for W3WAVE for the non-interpolated fields, but in this case an interpolated field must always be used when running WWIII (which does not seem unreasonable to me).

I managed to make this configuration work by initializing TFN(1) to zero instead of -1 in w3idatmd.ftn, but as there are various checks in the code for TFN(1) = -1 I am not sure this is a prudent thing to do. The rest of the changes I tried to make it work were unsuccessful.

**To Reproduce**
Run a configuration using water levels read from file, and winds or momentum as coupled field. 

**Expected behavior**
OASIS will crash because the order in which the coupling fields are exchanged is not right.

**Screenshots**

**Additional context**

### PR DESCRIPTION
**Pull request by ukmo-juan-castillo**
# Pull Request Summary
Bug fix for a problem when happens in certain circumstances (using driving data coming from oasis or from file in certain combinations), when oasis_get is called twice for the same field and time, causing OASIS to crash

## Description
It was discovered that, when reading water levels from file and driving the wave model from winds received via coupling, and coupling at T+0, the oasis_get call for winds at time start zero happens twice, which causes OASIS to crash because as expected the model that sends the winds only sends the fields once at each coupling step.

This is caused by a series of things: water levels are read before winds, and they are not time interpolated. This causes that the time step is not updated the first time the input fields are updated, so the input fields are updated twice the first time step.

The fix here proposed makes sure that oasis_get is called only once if the input fields are updated twice for the first time step.

To reproduce this error, I modified one ww3_tp2.14 test. The reviewer is encouraged to run the develop branch with the modified regtest to demonstrate that a bug is actually present, and that the changes in this branch fixes it. Given that this particular regtest is changed, it is expected to cause a change in answers in just that test.

Labels that should be added: _bug_

### Issue(s) addressed
- fixes #816 
- fixes noaa-emc/ww3/issues/816

### Commit Message
Fix duplicated calls to oasis_get happening in some mixed forced/coupled configurations

### Check list  

- [YES ] Branch is up to date with the authoritative repository (NOAA-EMC) develop branch. 
- [ YES] Checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop). 
- [ N/A] If a version number update is required, checked the [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number) checklist. 
- [YES ] If a new feature was added, a regression test for testing the new feature is added. 
 
### Testing

* How were these changes tested?
* Are the changes covered by regression tests? (If not, why? Do new tests need to be added?)
* Have the matrix regression tests been run (if yes, please note HPC and compiler)?
* Please indicate the expected changes in the regression test output, (Note the [list](https://github.com/NOAA-EMC/WW3/wiki/How-to-use-matrix.comp-to-compare-regtests-with-develop#4-look-at-results) of known non-identical tests.)
* Please provide the summary output of matrix.comp (_matrix.Diff.txt_, _matrixCompFull.txt_ and _matrixCompSummary.txt_):

### CODE REVIEW

**Comment by ukmo-ccbunney** [comment]
Hi @ukmo-juan-castillo - did you run the ww3_tp2.14 regtest with your fix prior to modifying the test? I just wanted to check that it was producing the same answers prior to your update.

I am going to run the full regtest suite today.

**Comment by JessicaMeixner-NOAA** [comment]
@ukmo-juan-castillo can you please merge in the noaa-emc/develop branch to your branch for this PR so that we can get the CI fixes here as well?  Thanks!

**Comment by ukmo-ccbunney** [comment]
> @ukmo-juan-castillo can you please merge in the noaa-emc/develop branch to your branch for this PR so that we can get the CI fixes here as well? Thanks!

I've done this now @JessicaMeixner-NOAA .

**Comment by ukmo-ccbunney** [comment]
@ukmo-juan-castillo - I am running the regression tests for this change and the model is hanging on the new `ww3_tp2.14/OASACM6` regtest.

If I revert the change to that regtest (setting `AIR_DENSITY="F"`) it still hangs.

**Comment by ukmo-ccbunney** [comment]
@ukmo-juan-castillo I've just pushed a small correction to the `ww3_from_ftp.sh` script - there was an incomplete destination path for the new density.nc file copied for `input_oasacm6`.

File: regtests/bin/matrix.base
Comment (ukmo-ccbunney) [inline regtests/bin/matrix.base]: @ukmo-juan-castillo : 
Why was -C OASIS added here? Isn't OASACM3 an uncoupled run? I am getting build errors for this test.

**Review by ukmo-ccbunney — CHANGES_REQUESTED** [review]

File: regtests/bin/matrix.base
Comment (ukmo-juan-castillo) [reply regtests/bin/matrix.base]: You are right, -C is not necessary if we have an uncoupled run. I forgot this was the case for the OASACM3 test.

File: regtests/ww3_tp2.14/input_oasicm/namcouple
Comment (ukmo-ccbunney) [inline regtests/ww3_tp2.14/input_oasicm/namcouple:32]: The addition of the 0 here is causing some output differences in the `nout.000000` file for all regtests. 
Difference is the addition of the following warning:

```
        ***WARNING***               
 Nothing on input for time statistic
 Default value 0 will be used       
```
Do we want to keep it?

**Review by ukmo-ccbunney — CHANGES_REQUESTED** [review]
@ukmo-juan-castillo 
I've rerun the ww3_tp2.14 regression tests.
There are some differences in the `nout.00000` files in all tests (see specific remarks below).

The OASACM6 test obviously has more difference due to the addition of the density field.
However, there are a couple of issues.
  - Firstly, the `density.nc` field used as input to `ww3_prnc` is for a different region (Adriatic Sea) so the input density.ww3 is all zero.
  - Secondly, in the output file `ww3.201109.nc`, the wind speeds are all zero (maybe a result of the density being zero?), so the resulting wave field is also zero.

I think we will need to add a new density.nc file for this regtest. It could be a small and very coarse resolution file with homogeneous values just to provide an input.

File: regtests/ww3_tp2.14/input_oasicm/namcouple
Comment (ukmo-ccbunney) [reply regtests/ww3_tp2.14/input_oasicm/namcouple:32]: @ukmo-juan-castillo I created a density.nc file that covered the region and reran the test, however the coupled wind are still showing as zero in the output file.

However, the waves heights are now non-zero.

Comment (ukmo-juan-castillo) [reply regtests/ww3_tp2.14/input_oasicm/namcouple:32]: I suspected that the changes in the namcouple files were the cause of the differences in the nout files. The reason of the cahnge is that the NLOG parameter in the namcouple file takes two integer values: the first is the debug level, and the second the type of timing information. If no time parameter is given (as it was originally the case), the default value is zero. The only thing I did was explicitly adding the default zero value to avoid possible problems when the coupler read this set of parameters.

I do not know why you have zero winds, I can check the output to see what happens.

Thanks for creating the new density file, another thing I overlooked... do not try to finish a ticket while you are in a workshop!

Comment (ukmo-juan-castillo) [reply regtests/ww3_tp2.14/input_oasicm/namcouple:32]: The problem with the winds was that the regtest is using the ww3_shel.inp file, and this file was not properly set. Instead of having an empty list of sent coupled fields, one needs to specify 'UNSET'.

Comment (ukmo-ccbunney) [reply regtests/ww3_tp2.14/input_oasicm/namcouple:32]: OK - results look ok to me now.
I am running the full regression test matrix.

@MatthewMasarik-NOAA and @JessicaMeixner-NOAA : We will need to add a new `density.nc` file to the `ww3_from_ftp.${ww3ver}.tar.gz` tar file for regtest `ww3_tp2.14/OASACM6`. Should we send this to you? It is only a small file (21Kb)

Comment (MatthewMasarik-NOAA) [reply regtests/ww3_tp2.14/input_oasicm/namcouple:32]: Hi @ukmo-ccbunney, yes that would be great.  Please send it to me and we will update it on the FTP site.

**Comment by ukmo-ccbunney** [comment]
I have run the full regression test matrix and get the following expected differences to `mww3_test_02` 
```
ww3_tp2.14/./work_OASACM                     (2 files differ)
ww3_tp2.14/./work_OASACM6                     (10 files differ)
ww3_tp2.14/./work_OASACM2                     (2 files differ)
ww3_tp2.14/./work_OASACM5                     (2 files differ)
ww3_tp2.14/./work_OASACM4                     (2 files differ)
ww3_tp2.14/./work_OASOCM                     (2 files differ)
ww3_tp2.14/./work_OASICM                     (2 files differ)
```

but there are also a few unexpected differences that remain even after rerunning:

 - `mww3_test_02/./work_PR3_UQ_MPI_c_c` - 2 log files differ (they are oddly empty in results from this branch)
 - `mww3_test_07/./work_PR3_UQ ` - mod_def file is different

I am a bit suspicious that this is some oddity to our system, so I wait to see what the results from NOAA's regteests are . Certainly it makes no sense that the `mod_def` file should be different in mww3_test_07 as Juan has now changed anything that could have affected ww3_grid.

File: regtests/ww3_tp2.14/input_oasicm/namcouple
Comment (ukmo-ccbunney) [reply regtests/ww3_tp2.14/input_oasicm/namcouple:32]: > Hi @ukmo-ccbunney, yes that would be great. Please send it to me and we will update it on the FTP site.

Thanks @MatthewMasarik-NOAA  - I've emailed you the new file.

Comment (MatthewMasarik-NOAA) [reply regtests/ww3_tp2.14/input_oasicm/namcouple:32]: Received.  Thank you @ukmo-ccbunney.

**Comment by MatthewMasarik-NOAA** [comment]
> I am a bit suspicious that this is some oddity to our system, so I wait to see what the results from NOAA's regteests are . Certainly it makes no sense that the `mod_def` file should be different in mww3_test_07 as Juan has now changed anything that could have affected ww3_grid.

@ukmo-ccbunney I've submitted the regtests for this.  We will report back on the differences you reported.

**Comment by JessicaMeixner-NOAA** [comment]
@ukmo-ccbunney can you update the version number to 7.14.1 instead of 7.12.7?   The density.nc file should be there now in that file.  Also, would you mind merging with the latest noaa-emc/develop?   I'm starting regtests now (Matt is on leave).

**Comment by ukmo-ccbunney** [comment]
> @ukmo-ccbunney can you update the version number to 7.14.1 instead of 7.12.7? The density.nc file should be there now in that file. Also, would you mind merging with the latest noaa-emc/develop? I'm starting regtests now (Matt is on leave).

Done! (online via github.dev, which was very exciting!)

**Comment by JessicaMeixner-NOAA** [comment]
@ukmo-ccbunney I ran the regtests yesterday and I didn't get either of the issues you mentioned.  I do get mod_def diffs for unstructured grids on a regular basis (although not for tp2.7 oddly enough), so I think that's just the known issue that we really need to get fixed.  I did just go and check the mww3_test_02/./work_PR3_UQ_MPI_c_c/ww3_multi.out file and it looks like we have some silent failures in that test that also exist in the develop branch.  I'm getting: " integration stalled: num_subseg exceeded limit"  
So perhaps something to look into separately but not something that should hold this PR up.  

I'm going to re-run the regtests today just to do a sanity check on the tar-file, but otherwise think this will be ready to merge in this afternoon unless we need to wait for @mickaelaccensi to review and approve as well.

**Review by JessicaMeixner-NOAA — APPROVED** [review]
```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR1_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (16 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (12 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (9 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (16 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (16 files differ)
mww3_test_03/./work_PR2_UQ_MPI_d2                     (14 files differ)
mww3_test_03/./work_PR3_UQ_MPI_e                     (1 files differ)
mww3_test_03/./work_PR3_UNO_MPI_e_c                     (1 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (16 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (7 files differ)
ww3_tp2.14/./work_OASACM6                     (10 files differ)
ww3_tp2.14/./work_OASACM5                     (2 files differ)
ww3_tp2.14/./work_OASACM4                     (2 files differ)
ww3_tp2.14/./work_OASACM                     (2 files differ)
ww3_tp2.14/./work_OASICM                     (2 files differ)
ww3_tp2.14/./work_OASACM2                     (2 files differ)
ww3_tp2.14/./work_OASOCM                     (2 files differ)
ww3_tp2.16/./work_MPI_OMPH                     (4 files differ)
ww3_tp2.17/./work_ma                     (1 files differ)
ww3_tp2.17/./work_ma1                     (1 files differ)
ww3_ufs1.3/./work_a                     (3 files differ)
```
[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/10547268/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/10547269/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/10547270/matrixDiff.txt)

**Comment by JessicaMeixner-NOAA** [comment]
@ukmo-ccbunney apologies for the slowness on the regtests on our end.  The second time I ran this branch, I had a few " 0 diffs" but looking into a case or two, it doesn't seem to be anything concerning.  @MatthewMasarik-NOAA or I will follow up with the mww3_test_02/./work_PR3_UQ_MPI_c_c test and make an issue to document this regression tests with the silent failure on our end and actual error on yours.    

@mickaelaccensi --- if you'd like to continue to review this PR, please do and if you find any issues or have any concerns we'll address them when you raise them.

**Comment by ukmo-ccbunney** [comment]
> @ukmo-ccbunney apologies for the slowness on the regtests on our end. The second time I ran this branch, I had a few " 0 diffs" but looking into a case or two, it doesn't seem to be anything concerning. @MatthewMasarik-NOAA or I will follow up with the mww3_test_02/./work_PR3_UQ_MPI_c_c test and make an issue to document this regression tests with the silent failure on our end and actual error on yours.
> 
> @mickaelaccensi --- if you'd like to continue to review this PR, please do and if you find any issues or have any concerns we'll address them when you raise them.

Great - Thanks Jessica.
Regarding the "0 diffs", I've had this before a few times
I think it is down to the existence or non-existence of `ww3_prep.out` files from the multigrid tests (if I remember rightly).

**Pull request merged**
